File: gamepack/Decompiled/NodeDeque.java
public class NodeDeque {
   Node an;
   public Node af = new Node();

   public NodeDeque() {
      this.af.hc = this.af;
      this.af.hg = this.af;
   }

   public void af() {
      while(true) {
         Node var1 = this.af.hc;
         if (var1 == this.af) {
            this.an = null;
            return;
         }

         var1.remove();
      }
   }

   public void addFirst(Node var1) {
      if (var1.hg != null) {
         var1.remove();
      }

      var1.hg = this.af.hg;
      var1.hc = this.af;
      var1.hg.hc = var1;
      var1.hc.hg = var1;
   }

   public void addLast(Node var1) {
      if (var1.hg != null) {
         var1.remove();
      }

      var1.hg = this.af;
      var1.hc = this.af.hc;
      var1.hg.hc = var1;
      var1.hc.hg = var1;
   }

   public static void ac(Node var0, Node var1) {
      if (var0.hg != null) {
         var0.remove();
      }

      var0.hg = var1.hg;
      var0.hc = var1;
      var0.hg.hc = var0;
      var0.hc.hg = var0;
   }

   public Node removeLast() {
      Node var1 = this.af.hc;
      if (var1 == this.af) {
         return null;
      } else {
         var1.remove();
         return var1;
      }
   }

   public Node removeFirst() {
      Node var1 = this.af.hg;
      if (var1 == this.af) {
         return null;
      } else {
         var1.remove();
         return var1;
      }
   }

   public Node last() {
      Node var1 = this.af.hc;
      if (var1 == this.af) {
         this.an = null;
         return null;
      } else {
         this.an = var1.hc;
         return var1;
      }
   }

   public Node first() {
      Node var1 = this.af.hg;
      if (var1 == this.af) {
         this.an = null;
         return null;
      } else {
         this.an = var1.hg;
         return var1;
      }
   }

   public Node previous() {
      Node var1 = this.an;
      if (var1 == this.af) {
         this.an = null;
         return null;
      } else {
         this.an = var1.hc;
         return var1;
      }
   }

   public Node next() {
      Node var1 = this.an;
      if (var1 == this.af) {
         this.an = null;
         return null;
      } else {
         this.an = var1.hg;
         return var1;
      }
   }
}

File: updater-runestar/src/main/java/org/runestar/client/updater/mapper/identifiers/classes/NodeDeque.kt
package org.runestar.client.updater.mapper.identifiers.classes

import org.objectweb.asm.Opcodes
import org.objectweb.asm.Type
import org.runestar.client.common.startsWith
import org.runestar.client.updater.mapper.abstractclasses.IdentityMapper
import org.runestar.client.updater.mapper.annotations.DependsOn
import org.runestar.client.updater.mapper.annotations.MethodParameters
import org.runestar.client.updater.mapper.predicateutilities.and
import org.runestar.client.updater.mapper.predicateutilities.predicateOf
import org.runestar.client.updater.mapper.predicateutilities.type
import org.runestar.client.updater.mapper.wrappers.Class2
import org.runestar.client.updater.mapper.wrappers.Field2
import org.runestar.client.updater.mapper.wrappers.Method2
import java.lang.reflect.Modifier

@DependsOn(Node::class)
class NodeDeque : IdentityMapper.Class() {
    override val predicate = predicateOf<Class2> { it.superType == Any::class.type }
            .and { it.instanceFields.size == 2 }
            .and { it.interfaces.isEmpty() }
            .and { it.instanceFields.count { it.type == type<Node>() } == 2 }

    @DependsOn(Node::class)
    class sentinel : IdentityMapper.InstanceField() {
        override val predicate = predicateOf<Field2> { it.type == type<Node>() }
                .and { Modifier.isPublic(it.access) }
    }

    @DependsOn(Node::class, sentinel::class)
    class current : IdentityMapper.InstanceField() {
        override val predicate = predicateOf<Field2> { it.type == type<Node>() }
                .and { it.id != field<sentinel>().id }
    }

    @MethodParameters("node")
    @DependsOn(Node::class, Node.next::class)
    class addFirst : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<Method2> { it.arguments.startsWith(type<Node>()) }
                .and { it.returnType == Type.VOID_TYPE }
                .and { it.instructions.filter { it.isField && it.fieldId == field<Node.next>().id }.count() == 5 }
    }

    @MethodParameters("node")
    @DependsOn(Node::class, Node.next::class)
    class addLast : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<Method2> { it.arguments.startsWith(type<Node>()) }
                .and { it.returnType == Type.VOID_TYPE }
                .and { it.instructions.filter { it.isField && it.fieldId == field<Node.next>().id }.count() == 4 }
    }

//    @MethodParameters
//    class clear : IdentityMapper.InstanceMethod() {
//        override val predicate = predicateOf<Method2> { it.arguments.size in 0..1 }
//                .and { it.returnType == Type.VOID_TYPE }
//                .and { it.instructions.any { it.opcode == Opcodes.GOTO } }
//    }

    @MethodParameters
    @DependsOn(Node::class, Node.remove::class, Node.next::class)
    class next : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<Method2> { it.arguments.size in 0..1 }
                .and { it.returnType == type<Node>() }
                .and { it.instructions.none { it.isMethod && it.methodId == method<Node.remove>().id } }
                .and { it.instructions.count { it.isField && it.fieldId == field<Node.next>().id } == 1 }
    }

    @MethodParameters
    @DependsOn(Node::class, Node.remove::class, Node.previous::class)
    class previous : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<Method2> { it.arguments.size in 0..1 }
                .and { it.returnType == type<Node>() }
                .and { it.instructions.none { it.isMethod && it.methodId == method<Node.remove>().id } }
                .and { it.instructions.count { it.isField && it.fieldId == field<Node.previous>().id } == 1 }
    }

    @MethodParameters
    @DependsOn(Node::class, Node.remove::class, Node.next::class)
    class first : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<Method2> { it.arguments.size in 0..1 }
                .and { it.returnType == type<Node>() }
                .and { it.instructions.none { it.isMethod && it.methodId == method<Node.remove>().id } }
                .and { it.instructions.count { it.isField && it.fieldId == field<Node.next>().id } == 2 }
    }

    @MethodParameters
    @DependsOn(Node::class, Node.remove::class, Node.previous::class)
    class last : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<Method2> { it.arguments.size in 0..1 }
                .and { it.returnType == type<Node>() }
                .and { it.instructions.none { it.isMethod && it.methodId == method<Node.remove>().id } }
                .and { it.instructions.count { it.isField && it.fieldId == field<Node.previous>().id } == 2 }
    }

    @MethodParameters
    @DependsOn(Node::class, Node.remove::class, Node.next::class)
    class removeFirst : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<Method2> { it.arguments.size in 0..1 }
                .and { it.returnType == type<Node>() }
                .and { it.instructions.any { it.isMethod && it.methodId == method<Node.remove>().id } }
                .and { it.instructions.count { it.isField && it.fieldId == field<Node.next>().id } == 1 }
    }

    @MethodParameters
    @DependsOn(Node::class, Node.remove::class, Node.previous::class)
    class removeLast : IdentityMapper.InstanceMethod() {
        override val predicate = predicateOf<Method2> { it.arguments.size in 0..1 }
                .and { it.returnType == type<Node>() }
                .and { it.instructions.any { it.isMethod && it.methodId == method<Node.remove>().id } }
                .and { it.instructions.count { it.isField && it.fieldId == field<Node.previous>().id } == 1 }
    }
}